blog/2025-01-17-how-are-the-messages-stored/index.md

```diff
@@ -0,0 +1,203 @@
+stored messages on disk and didn't keep them in memory at all. This means that "RabbitMQ stores messages in memory"
+was not true 10 years ago. It'd still do this by default, but it was completely optional.
+
+Lazy mode was removed in RabbitMQ in 3.12, released in early 2024, but the default (and the only available) behaviour
+changed and is similar to the lazy mode, although not exactly the same. Therefore, for a year now, classic queues don't store
+messages in memory and can't even be configured to do so. The myth is almost entirely false at this point.
+
+Almost entirely? So here's how classic queues work right now: they accumulate incoming messages in a small in-memory buffer
+and write them to disk as a batch, as soon as the in-memory buffer is full. Since we don't know if more messages will come,
+there are additional triggers that flush that buffer, including a flush after a certain number of batched messages
+(even if they are too small to use the whole buffer) and after
+a certain number of operations on that batch is performed. Additionally, the queue monitors how quickly messages are consumed
+and makes decisions based on that (if there are fast consumers, more messages will be cached in memory). Finally, there's the
+last-resort trigger which will flush the buffer every 200ms. So the absolute maximum of how long a message can be stored in
+memory is 200ms but in practice, I've never seen this happen. Publishers usually receive confirms within a few milliseconds
+and they are sent only after the message was written to disk.
+
+### But I See No Disk Activity When I Use a Classic Queue!
+
+Indeed, it's entirely possible to publish messages to classic queues and see virtually no disk writes nor disk reads.
+How is that possible? It's an optimisation for a very specific, but relatively common case. As explained above, messages
+can be briefly kept in memory but if there are active consumers that are waiting for messages (their prefetch buffer is not full),
+messages that reach the queue are dispatched to the consumers immediately, without waiting for their batch to be written to disk.
+A message that gets acknowledged by the consumer before this message's batch is written to disk, will not be written to disk at all,
+because it simply doesn't need to be. Queues don't store acknowledged messages, so if the message is acknowledged before it is written,
+it doesn't get written. If it's acknowledged after it was written, it's deleted from the queue (the actual removal from disk
+will happen later, asynchronously, but it is considered deleted immediately).
+
+It's worth mentioning that classic queues have [two separate storage mechanisms](https://www.rabbitmq.com/blog/2024/01/11/3.13-release#classic-queues-storage-primer).
+Messages below 4kb (configurable through `queue_index_embed_msgs_below`) are stored in a per-queue message store and message above that threshold,
```

Comment:
"message above that threshold," => "messages above that threshold" (missing 's' and extra comma).